components/script/dom/globalscope.rs

```diff
@@ -2084,6 +2091,26 @@ impl GlobalScope {
         unreachable!();
     }
 
+    // https://w3c.github.io/performance-timeline/#supportedentrytypes-attribute
+    pub fn supported_performance_entry_types(&self, cx: SafeJSContext) -> JSVal {
+        if let Some(types) = &*self.frozen_supported_performance_entry_types.borrow() {
+            return types.get();
+        }
+
+        let types: Vec<DOMString> = VALID_ENTRY_TYPES
+            .iter()
+            .map(|t| DOMString::from(t.to_string()))
+            .collect();
+        let frozen_types = to_frozen_array(types.as_slice(), cx);
+
+        // Cache the Js value.
+        let heap_val = Heap::default();
+        heap_val.set(frozen_types);
+        *self.frozen_supported_performance_entry_types.borrow_mut() = Some(heap_val);
```

Comment:
This way of using Heap values in Option values isn't actually safe; I filed #25726 about finding a way to avoid this problem. In the short term, please use:
```rust
// Safety: need to create the Heap value in its final memory location before setting it.
*self.frozen_supported_performance_entry_types.borrow_mut() = Some(Heap::default());
self.frozen_supported_performance_entry_types.borrow().as_ref().unwrap().set(frozen_types);
```

Reply:
That came straight from MessageEvent::Ports; I'll fix it there too.